Add protocol and configuration bindings for outbound SNS.

TreyE opened the pull request:
This adds protocol and configuration bindings for outbound SNS messaging using AWS, particularly for SMS messages.

Testing is still pending, as a working application generating events is needed to verify the implementation.

https://app.clickup.com/t/868d7e750

TreyE committed:
Add protocol and configuration bindings for outbound SNS.

This adds protocol and configuration bindings for outbound SNS messaging
using AWS, particularly for SMS messages.

Testing is still pending, as a working application generating events is
needed to verify the implementation.

raghuramg commented on lib/event_source/protocols/base/connection_proxy.rb:
```diff
@@ -0,0 +1,38 @@
+  module Protocols
+    module Base
+      # Connection proxy common behaviour.
+      class ConnectionProxy
```
It looks like this file isn't being used anywhere. Do we plan to use it in the future? If not, can we remove it?

TreyE replied:
It isn't being used currently except as a base interface for the purposes of documentation when we implement new protocols - and given that isn't clear from the PR - it needs to be more clear why we have a file that as you say 'isn't used'.

raghuramg commented on lib/event_source/protocols/sns/publish_bindings.rb:
```diff
@@ -0,0 +1,23 @@
+        attribute :protocol, Types::String.meta(omittable: false)
+
+        def sms?
+          protocol.to_s.downcase == "sms"
```
I noticed that in lib/event_source/protocols/sns/contracts/publish_operation_binding_contract.rb, we're expecting sms as the protocol. Is sns considered the protocol and sms the message type? Could you clarify the abstraction being used here?

raghuramg commented on lib/event_source/protocols/registry.rb:
```diff
@@ -0,0 +1,33 @@
+module EventSource
+  module Protocols
+    # Registers protocols by URI format for later lookup.
+    class Registry
```
Can we add a spec to test the registration and lookup of a protocol?

TreyE replied:
I am of the opinion that this is accurately covered using the inner workings of event source - but adding a new unit test to enforce the interface and it's usage is a good idea - will do.

raghuramg reviewed (changes requested):
I've added a few comments—could you please review them?

In addition to the new protocol implementation, we also improved error messaging, introduced a base channel proxy, and cleaned up the existing operations. I'm happy with these changes.

TreyE commented:
> I've added a few comments—could you please review them?
> 
> In addition to the new protocol implementation, we also improved error messaging, introduced a base channel proxy, and cleaned up the existing operations. I'm happy with these changes.

Absolutely, basic gist of the comments (which I agree with whole heartedly) is the areas in which you have questions I don't provide sufficient documentation or explain what/why I'm doing what is happening in the code.  I will do so before I re-submit.

raghuramg reviewed (approved):
Looks good to me

polographer reviewed (approved):
My only comment is that I didn't see a re-connection explicitly in the code. I guess that is to keep the current behaviour of the gem; however, for some reason, I think this new kind of connection is more prone to disconnections. But overall, great work
